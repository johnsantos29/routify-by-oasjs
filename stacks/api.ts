import { SecretsStack } from "./secrets";
import { use, Stack, StackContext, Api, Function } from "@serverless-stack/resources";
import { CustomDomain, ApiHandlers } from "../types/routify";

const permissions = ["events", "ssm", "rds-data"];

// api construct
const routifyApi = (stack: Stack, handlers: ApiHandlers, domain: CustomDomain) =>
    new Api(stack, "routifyApi", {
        customDomain: {
            domainName: domain.domainName,
            hostedZone: domain.hostedZone,
        },
        routes: {
            "GET /home-time": handlers.homeTimeHandler,
        },
    });

export const APIStack = ({ app, stack }: StackContext) => {
    // const { vpcId } = use(SecretsStack);

    // lambdas
    const homeTimeHandler = new Function(stack, "homeTime", {
        architecture: "arm_64",
        memorySize: "256 MB",
        runtime: "nodejs16.x",
        handler: "functions/home-time.handler",
        permissions,
    });

    const domain = {
        domainName: "api.oasjs.com",
        hostedZone: "oasjs.com",
    };

    // http api
    const api = routifyApi(stack, { homeTimeHandler }, domain);

    // show the endpoint in the output
    stack.addOutputs({
        ApiEndpoint: api.url,
    });
};
