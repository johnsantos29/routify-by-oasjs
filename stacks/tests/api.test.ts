import { test, expect } from "vitest";
import { Template } from "aws-cdk-lib/assertions";
import { App, getStack } from "@serverless-stack/resources";
import { APIStack } from "../api";

test("the creation of API Stack", () => {
    const app = new App();
    app.stack(APIStack);

    // get template

    const apiStack = getStack(APIStack);
    const apiTemplate = Template.fromStack(apiStack);

    expect(apiTemplate.toJSON()).toMatchSnapshot();
});
