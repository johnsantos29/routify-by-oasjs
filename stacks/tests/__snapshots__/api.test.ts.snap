// Vitest Snapshot v1

exports[`the creation of API Stack 1`] = `
{
  "Outputs": {
    "ApiEndpoint": {
      "Value": {
        "Fn::GetAtt": [
          "routifyApiC0F7BBB8",
          "ApiEndpoint",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomResourceHandlerE8FB56BA": {
      "DependsOn": [
        "CustomResourceHandlerServiceRoleDefaultPolicy73F93F56",
        "CustomResourceHandlerServiceRole41AEC181",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-my-account-us-east-1",
          "S3Key": "ad85de0764d9cc133f45a9f81e4a652509a0bfa17ec6c748e7953e5f125f3172.zip",
        },
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "CustomResourceHandlerServiceRole41AEC181",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Tags": [
          {
            "Key": "sst:app",
            "Value": "my-app",
          },
          {
            "Key": "sst:stage",
            "Value": "dev",
          },
        ],
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomResourceHandlerServiceRole41AEC181": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "sst:app",
            "Value": "my-app",
          },
          {
            "Key": "sst:stage",
            "Value": "dev",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomResourceHandlerServiceRoleDefaultPolicy73F93F56": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutObject",
                "s3:DeleteObject",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::undefined/*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomResourceHandlerServiceRoleDefaultPolicy73F93F56",
        "Roles": [
          {
            "Ref": "CustomResourceHandlerServiceRole41AEC181",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StackMetadata": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "App": "my-app",
        "Metadata": [
          {
            "addr": "c864e26ca4c028578a261425a262d3c60150ef58be",
            "data": {
              "arn": {
                "Fn::GetAtt": [
                  "homeTime913CF0F4",
                  "Arn",
                ],
              },
              "localId": "dev-my-app-APIStack-homeTime",
              "secrets": [],
            },
            "id": "homeTime",
            "stack": "dev-my-app-APIStack",
            "type": "Function",
          },
          {
            "addr": "c876ae46cbe03de0a175fc13bf05f9027933c5754e",
            "data": {
              "customDomainUrl": "https://api.oasjs.com",
              "graphql": false,
              "httpApiId": {
                "Ref": "routifyApiC0F7BBB8",
              },
              "routes": [
                {
                  "fn": {
                    "node": "c864e26ca4c028578a261425a262d3c60150ef58be",
                    "stack": "dev-my-app-APIStack",
                  },
                  "route": "GET /home-time",
                  "type": "function",
                },
              ],
              "url": {
                "Fn::Join": [
                  "",
                  [
                    "https://",
                    {
                      "Ref": "routifyApiC0F7BBB8",
                    },
                    ".execute-api.us-east-1.",
                    {
                      "Ref": "AWS::URLSuffix",
                    },
                    "/",
                  ],
                ],
              },
            },
            "id": "routifyApi",
            "stack": "dev-my-app-APIStack",
            "type": "Api",
          },
        ],
        "SSTVersion": "1.15.8",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomResourceHandlerE8FB56BA",
            "Arn",
          ],
        },
        "Stack": "dev-my-app-APIStack",
        "Stage": "dev",
      },
      "Type": "Custom::StackMetadata",
      "UpdateReplacePolicy": "Delete",
    },
    "homeTime913CF0F4": {
      "DependsOn": [
        "homeTimeServiceRoleDefaultPolicy7F40DCE9",
        "homeTimeServiceRole6296CFCF",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "ZipFile": "export function placeholder() {}",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SST_APP": "my-app",
            "SST_STAGE": "dev",
          },
        },
        "EphemeralStorage": {
          "Size": 512,
        },
        "Handler": "index.placeholder",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "homeTimeServiceRole6296CFCF",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Tags": [
          {
            "Key": "sst:app",
            "Value": "my-app",
          },
          {
            "Key": "sst:stage",
            "Value": "dev",
          },
        ],
        "Timeout": 10,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "homeTimeServiceRole6296CFCF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "sst:app",
            "Value": "my-app",
          },
          {
            "Key": "sst:stage",
            "Value": "dev",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "homeTimeServiceRoleDefaultPolicy7F40DCE9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "events:*",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "ssm:*",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "rds-data:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "homeTimeServiceRoleDefaultPolicy7F40DCE9",
        "Roles": [
          {
            "Ref": "homeTimeServiceRole6296CFCF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "routifyApiAliasRecord386FC185": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "routifyApiDomainNameCE6B535D",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::GetAtt": [
              "routifyApiDomainNameCE6B535D",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "api.oasjs.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "routifyApiAliasRecordAAAA017E2E03": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "routifyApiDomainNameCE6B535D",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::GetAtt": [
              "routifyApiDomainNameCE6B535D",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "api.oasjs.com.",
        "Type": "AAAA",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "routifyApiC0F7BBB8": {
      "Properties": {
        "CorsConfiguration": {
          "AllowCredentials": false,
          "AllowHeaders": [
            "*",
          ],
          "AllowMethods": [
            "*",
          ],
          "AllowOrigins": [
            "*",
          ],
        },
        "Name": "dev-my-app-routifyApi",
        "ProtocolType": "HTTP",
        "Tags": {
          "sst:app": "my-app",
          "sst:stage": "dev",
        },
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "routifyApiCertificate08D502B1": {
      "Properties": {
        "DomainName": "api.oasjs.com",
        "DomainValidationOptions": [
          {
            "DomainName": "api.oasjs.com",
            "HostedZoneId": "DUMMY",
          },
        ],
        "Tags": [
          {
            "Key": "sst:app",
            "Value": "my-app",
          },
          {
            "Key": "sst:stage",
            "Value": "dev",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "routifyApiDefaultStage7C2547EE": {
      "DependsOn": [
        "routifyApiDomainNameCE6B535D",
      ],
      "Properties": {
        "AccessLogSettings": {
          "DestinationArn": {
            "Fn::GetAtt": [
              "routifyApiLogGroupC476DC9B",
              "Arn",
            ],
          },
          "Format": "{\\"requestTime\\":\\"$context.requestTime\\",\\"requestId\\":\\"$context.requestId\\",\\"httpMethod\\":\\"$context.httpMethod\\",\\"path\\":\\"$context.path\\",\\"routeKey\\":\\"$context.routeKey\\",\\"status\\":$context.status,\\"responseLatency\\":$context.responseLatency,\\"integrationRequestId\\":\\"$context.integration.requestId\\",\\"integrationStatus\\":\\"$context.integration.status\\",\\"integrationLatency\\":\\"$context.integration.latency\\",\\"integrationServiceStatus\\":\\"$context.integration.integrationStatus\\",\\"ip\\":\\"$context.identity.sourceIp\\",\\"userAgent\\":\\"$context.identity.userAgent\\",\\"cognitoIdentityId\\":\\"$context.identity.cognitoIdentityId\\"}",
        },
        "ApiId": {
          "Ref": "routifyApiC0F7BBB8",
        },
        "AutoDeploy": true,
        "StageName": "$default",
        "Tags": {
          "sst:app": "my-app",
          "sst:stage": "dev",
        },
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "routifyApiDefaultStagedevmyappAPIStackroutifyApiDomainNameundefinedB1056927": {
      "DependsOn": [
        "routifyApiDefaultStage7C2547EE",
        "routifyApiDomainNameCE6B535D",
      ],
      "Properties": {
        "ApiId": {
          "Ref": "routifyApiC0F7BBB8",
        },
        "DomainName": {
          "Ref": "routifyApiDomainNameCE6B535D",
        },
        "Stage": "$default",
      },
      "Type": "AWS::ApiGatewayV2::ApiMapping",
    },
    "routifyApiDomainNameCE6B535D": {
      "Properties": {
        "DomainName": "api.oasjs.com",
        "DomainNameConfigurations": [
          {
            "CertificateArn": {
              "Ref": "routifyApiCertificate08D502B1",
            },
            "EndpointType": "REGIONAL",
          },
        ],
        "Tags": {
          "sst:app": "my-app",
          "sst:stage": "dev",
        },
      },
      "Type": "AWS::ApiGatewayV2::DomainName",
    },
    "routifyApiLogGroupC476DC9B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/vendedlogs/apis/dev-my-app-routifyApi-",
              {
                "Ref": "routifyApiC0F7BBB8",
              },
              "/default",
            ],
          ],
        },
        "Tags": [
          {
            "Key": "sst:app",
            "Value": "my-app",
          },
          {
            "Key": "sst:stage",
            "Value": "dev",
          },
        ],
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "routifyApiRouteGEThometime4F730C0A": {
      "Properties": {
        "ApiId": {
          "Ref": "routifyApiC0F7BBB8",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "GET /home-time",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "routifyApiRouteGEThometimeIntegrationGEThometimeE3B88CCF",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "routifyApiRouteGEThometimeIntegrationGEThometimeE3B88CCF": {
      "Properties": {
        "ApiId": {
          "Ref": "routifyApiC0F7BBB8",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "homeTime913CF0F4",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "routifyApiRouteGEThometimeIntegrationGEThometimePermissionACE913D9": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "homeTime913CF0F4",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:my-account:",
              {
                "Ref": "routifyApiC0F7BBB8",
              },
              "/*/*/home-time",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
